### PR TITLE
nanoCLR available as .NET tool

### PR DESCRIPTION
**Pull request by josesimoes**
## Description
- Make nanoCLR available as .NET tool.
- Make nanoCLR available as Windows DLL (and paving way for other platforms).

## Motivation and Context
- Improves usability for non-attended usage scenarios: pipelines, containers, test framework.
- Make way for the availability of "virtual" nano devices.

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [x] My code follows the code style of this project (only if there are changes in source code).
- [x] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

### CODE REVIEW

File: targets/netcore/nanoFramework.nanoCLR.Host/HexDump.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace HexDump
+{
+    public static class HexDump
+    {
+        [Pure]
+        public static string Format(byte[] bytes, int columnWidth = 8, int columnCount = 2, bool includeOffset = true, bool includeAscii = true)
+        {
+            if (bytes == null)
+                throw new ArgumentNullException(nameof(bytes));
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR.Host/HexDump.cs:17]: brackets ;-)

File: targets/netcore/nanoFramework.nanoCLR.Host/HexDump.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+            if (bytes.Length == 0)
+                return string.Empty;
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR.Host/HexDump.cs:20]: brackets ;-)

File: targets/netcore/nanoFramework.nanoCLR.CLI/RunCommandProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                    return -1;
+                }
+            }
+
+            if (options.Assemblies.Any())
+            {
+                hostBuilder.LoadAssemblies(options.Assemblies);
+            }
+
+            //if (options.AssembliesSet != null)
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR.CLI/RunCommandProcessor.cs:73]: some dead code here

File: targets/netcore/nanoFramework.nanoCLR.Host/Port/Channel.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
+//
+// Copyright (c) .NET Foundation and Contributors
+// See LICENSE file in the project root for full license information.
+//
+
+using System.Threading;
+
+namespace nanoFramework.nanoCLR.Host.Port
+{
+    public class Channel : IChannel
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR.Host/Port/Channel.cs:10]: Just a question on the visibility for all those classes and the interfaces, do they all need to be public or internal would be enough?

File: targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+                }
+                *start = (char *)datAssembliesBuffer;
+                *end = (char *)((unsigned int)datAssembliesBuffer + (unsigned int)datSize);
+            }
+        }
+
+        // else data in RAM
+        NANOCLR_NOCLEANUP();
+    }
+//
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp:222]: And some dead code here as well ;-)

File: targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp [hunk 6]
```diff
@@ -0,0 +1,670 @@
...
+        {
+            headerBuffer = (unsigned char *)CLR_RT_Memory::Allocate(headerInBytes, true);
+            CHECK_ALLOCATION(headerBuffer);
+            CLR_RT_Memory::ZeroFill(headerBuffer, headerInBytes);
+        }
+
+        while (TRUE)
+        {
+            if (!BlockStorageStream_Read(&stream, &headerBuffer, headerInBytes))
+                break;
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp:272]: explicit brackets

File: targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+                    CLR_RT_Memory::Release(headerBuffer);
+
+                    NANOCLR_SET_AND_LEAVE(CLR_E_OUT_OF_MEMORY);
+                }
+            }
+
+            BlockStorageStream_Seek(&stream, -headerInBytes, BlockStorageStream_SeekCurrent);
+
+            if (!BlockStorageStream_Read(&stream, &assembliesBuffer, AssemblySizeInByte))
+                break;
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp:303]: explicit brackets

File: targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp [hunk 8]
```diff
@@ -0,0 +1,670 @@
...
+            }
+
+            // we have good Assembly
+            CLR_RT_Assembly *assm;
+
+            // Creates instance of assembly, sets pointer to native functions, links to g_CLR_RT_TypeSystem
+            if (FAILED(LoadAssembly(header, assm)))
+            {
+                if (!isXIP)
+                    CLR_RT_Memory::Release(assembliesBuffer);
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp:321]: explicit brackets

File: targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp [hunk 9]
```diff
@@ -0,0 +1,670 @@
...
+
+            if (!BlockStorageStream_Read(&stream, &assembliesBuffer, AssemblySizeInByte))
+                break;
+
+            header = (const CLR_RECORD_ASSEMBLY *)assembliesBuffer;
+
+            if (!header->GoodAssembly())
+            {
+                if (!isXIP)
+                    CLR_RT_Memory::Release(assembliesBuffer);
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp:310]: explicit brackets

File: targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp [hunk 10]
```diff
@@ -0,0 +1,670 @@
...
+            if (FAILED(LoadAssembly(header, assm)))
+            {
+                if (!isXIP)
+                    CLR_RT_Memory::Release(assembliesBuffer);
+                break;
+            }
+            assm->m_flags |= CLR_RT_Assembly::Deployed;
+        }
+        if (!isXIP)
+            CLR_RT_Memory::Release(headerBuffer);
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp:327]: explicit brackets

File: targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp [hunk 11]
```diff
@@ -0,0 +1,670 @@
...
+                        src->version.iRevisionNumber);
+
+                    fError = true;
+                }
+            }
+        }
+        NANOCLR_FOREACH_ASSEMBLY_END();
+
+        if (fError)
+            NANOCLR_SET_AND_LEAVE(CLR_E_ENTRY_NOT_FOUND);
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp:555]: explicit brackets

File: targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp [hunk 12]
```diff
@@ -0,0 +1,670 @@
...
+        {
+            if (CLR_EE_REBOOT_IS(ClrOnly))
+            {
+                softReboot = true;
+
+                params.WaitForDebugger = CLR_EE_REBOOT_IS(WaitForDebugger);
+
+                s_ClrSettings.Cleanup();
+
+                // nanoHAL_Uninitialize();
```
Comment (Ellerbach) [inline targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp:659]: some dead code here, is it normal?

**Review by Ellerbach — COMMENTED** [review]
Minor comments and few questions. All this looks really good :-)

File: targets/netcore/nanoFramework.nanoCLR.Host/HexDump.cs (hunk 1 above)
Comment (josesimoes) [reply targets/netcore/nanoFramework.nanoCLR.Host/HexDump.cs:17]: All this source file was copied from the original source.

File: targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp (hunk 12 above)
Comment (josesimoes) [reply targets/netcore/nanoFramework.nanoCLR/CLRStartup.cpp:659]: Nice catch!

File: targets/netcore/nanoFramework.nanoCLR.Host/Port/Channel.cs (hunk 4 above)
Comment (josesimoes) [reply targets/netcore/nanoFramework.nanoCLR.Host/Port/Channel.cs:10]: Good point. Let me change that.

**Review by Ellerbach — APPROVED** [review]
That's a lot of code to deeply review, I did a quick check earlier and all up, all good. So approving this PR as in all cases, those are improvement on the nanoCLR that is already working and we can always adjust anything later on!

**Pull request merged**